invokeai/app/services/batch_manager.py

```diff
@@ -0,0 +1,206 @@
+        try:
+            next_session = self.__batch_process_storage.get_next_session(batch_id)
+        except BatchSessionNotFoundException:
+            return
+        batch_process = self.__batch_process_storage.get(batch_id)
+        ges = self._create_batch_session(batch_process=batch_process, batch_indices=tuple(next_session.batch_index))
+        ges.id = next_session.session_id
+        self.__invoker.services.graph_execution_manager.set(ges)
+        self.__batch_process_storage.update_session_state(
+            batch_id=next_session.batch_id,
+            session_id=next_session.session_id,
+            changes=BatchSessionChanges(state="in_progress"),
+        )
+        self.__invoker.invoke(ges, invoke_all=True)
+
+    def create_batch_process(self, batch: Batch, graph: Graph) -> BatchProcessResponse:
+        batch_process = BatchProcess(
+            batch=batch,
+            graph=graph,
+        )
+        batch_process = self.__batch_process_storage.save(batch_process)
+        sessions = self._create_sessions(batch_process)
+        return BatchProcessResponse(
+            batch_id=batch_process.batch_id,
+            session_ids=[session.session_id for session in sessions],
+        )
+
+    def _create_sessions(self, batch_process: BatchProcess) -> list[BatchSession]:
+        batch_indices = list()
+        sessions_to_create: list[BatchSession] = list()
+        for batchdata in batch_process.batch.data:
+            batch_indices.append(list(range(len(batchdata[0].items))))
+        all_batch_indices = product(*batch_indices)
+        for bi in all_batch_indices:
+            for _ in range(batch_process.batch.runs):
+                sessions_to_create.append(BatchSession(batch_id=batch_process.batch_id, batch_index=list(bi)))
+            if not sessions_to_create:
+                sessions_to_create.append(BatchSession(batch_id=batch_process.batch_id, batch_index=list(bi)))
+        created_sessions = self.__batch_process_storage.create_sessions(sessions_to_create)
+        return created_sessions
+
+    def get_sessions(self, batch_id: str) -> list[BatchSession]:
+        return self.__batch_process_storage.get_sessions_by_batch_id(batch_id)
+
+    def get_batch(self, batch_id: str) -> BatchProcess:
+        return self.__batch_process_storage.get(batch_id)
+
+    def get_batch_processes(self) -> list[BatchProcessResponse]:
+        bps = self.__batch_process_storage.get_all()
+        res = list()
+        for bp in bps:
+            sessions = self.__batch_process_storage.get_sessions_by_batch_id(bp.batch_id)
+            res.append(
+                BatchProcessResponse(
+                    batch_id=bp.batch_id,
+                    session_ids=[session.session_id for session in sessions],
+                )
+            )
+        return res
+
+    def get_incomplete_batch_processes(self) -> list[BatchProcessResponse]:
+        bps = self.__batch_process_storage.get_incomplete()
+        res = list()
```

Comment:
This "get list of sessions for batch" code seems to be repeated. Might want to extract that to a function.